rust/diesel/src/order.rs

```diff
@@ -0,0 +1,173 @@
+                .values(&row)
+                .get_result(connection)
+                .unwrap();
+
+            let order_lines: Vec<NewOrderLine> = product_units
+                .to_vec()
+                .into_iter()
+                .map(|(product, units)| NewOrderLine {
+                    order_id: inserted_order.order_id,
+                    product_id: product.product_id,
+                    units,
+                })
+                .collect();
+
+            diesel::insert_into(order_lines::table)
+                .values(order_lines)
+                .execute(connection)
+        });
+
+        insert_result.map(|_| order).ok()
+    }
+
+    pub fn read_all(connection: &PgConnection) -> Vec<Order> {
+        orders::table
+            .order(orders::order_id)
+            .load::<Order>(connection)
+            .unwrap()
+    }
+
+    pub fn read(id: i32, connection: &PgConnection) -> Option<Order> {
+        orders::table
+            .filter(orders::order_id.eq(id))
+            .first::<Order>(connection)
+            .ok()
+    }
+
+    pub fn update(id: i32, order: NewUserOrder, connection: &PgConnection) -> Option<NewUserOrder> {
+        let product_units: Vec<(Product, i16)> = order
+            .products
+            .iter()
+            .flat_map(|product_ref| {
+                Product::read(product_ref.product_id, connection)
+                    .map(|product| (product, product_ref.units))
+            })
+            .collect();
+
+        let order_total = product_units
+            .to_vec()
+            .into_iter()
+            .fold(BigDecimal::from(0.0), |acc, (product, qty)| {
+                acc.add(product.price.mul(BigDecimal::from(qty)))
+            });
+
+        if order_total.is_zero() {
+            return None;
+        }
+
+        let row = Order {
+            user_id: order.user_id,
+            order_id: id,
+            order_total,
+        };
+
+        let update_result = connection.build_transaction().read_write().run(|| {
+            let inserted_order: Order = diesel::update(orders::table.find(id))
+                .set(&row)
+                .get_result(connection)
+                .unwrap();
+
+            let order_lines: Vec<NewOrderLine> = product_units
+                .to_vec()
+                .into_iter()
+                .map(|(product, units)| NewOrderLine {
+                    order_id: inserted_order.order_id,
+                    product_id: product.product_id,
+                    units,
+                })
+                .collect();
+
+            diesel::insert_into(order_lines::table)
+                .values(order_lines)
+                .execute(connection)
+        });
+
+        update_result.map(|_| order).ok()
+    }
+
+    pub fn delete(id: i32, connection: &PgConnection) -> bool {
+        diesel::delete(orders::table.find(id))
+            .execute(connection)
+            .is_ok()
```

Comment:
I think this should return a `Result` instead of using a `bool` to represent the result.

Reply:
In this example, it is good enough to know if deletion succeeded or failed, we do not consume and act upon failure reason.